src/Library/Fsm/FSMState.h

```diff
@@ -0,0 +1,51 @@
+#pragma once
+
+#include <string_view>
+
+#include "FSMEventHandler.h"
+
+class FSMTransitionHandler;
+
+/**
+ * @brief Implement this abstract class to create a new State that can be added to an FSM.
+ * FSMState derives from FSMEventHandler, enabling reaction to platform application events (such as input, window events, etc.) 
+ * by overriding the appropriate functions.
+ */
+class FSMState: public FSMEventHandler {
+ public:
+    virtual ~FSMState() = default;
+
+    /*
+    * @brief Called during each FSM::update() iteration if this state is the current state in the FSM.
+    */
+    virtual void update() = 0;
+
+    /**
+     * @brief Invoked within FSM::update() when transitioning to this state as the new current state.
+     */
+    virtual void enter() = 0;
```

Comment:
I'd suggest a slightly different design:

* Drop `FSMTransitionHandler`, `setTransitionHandler`, `_scheduleTransition `.
* Add `FSMAction` abstracting away a transition request. For now you'll have either a non-action, or an action containing a target.
* Return `FSMAction` from both `enter` (for pass-through states) and `update`.

Reply:
I like it! It's actually enforcing even more the fact that transitions should happen inside those methods.

Reply:
Now `enter` and `update` returns a `FSMAction`. I went for an `std::variant` approach cause I didn't want to instantiate on the heap a Null object all the time ( I could use a static object but there would be other little things to solve in that case ). Maybe there's a better way but the two `FSMAction` we have are small enough to not be an issue. Of course, we can change it if more actions starts to be needed and they are big enough to invalidate the usage of a `std::variant`.

Reply:
Honestly what we have there is a bit of an overkill since you're doing packaging logic into `FSMAction `, and I don't think it's warranted there.

Realistically it's just transitions that are used as actions right now, so smth like this shall do:

```cpp
class FSMAction {
public:
    static FSMAction none() {
        return {};
    }
    
    static FSMAction transition(std::string_view target) {
        FSMAction result;
        result._target = target;
        return result;
    }

    operator bool() const {
        return !_target.empty();
    }

    bool operator!() const {
        return _target.empty();
    }

    const std::string &target() const {
        assert(!!*this);
        return _target;
    }

private:
    std::string _target;
}
```

And then pack all the logic that's processing the `FSMAction` inside the `FSM`.

Or maybe I'm missing smth. Let's discuss.

Reply:
Ahah, nvm I thought of something that simple but I misunderstood your "_abstracting away a transition request_" and I went down that path. I can change into what you're proposing.